### PR TITLE
Add local to global index mapping

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR adds a local-to-global mapping to the `index_map` class. It is necessary for the follow-up PR #1639. This also introduces a device-side view of a partition, and adds another member to the partition which stores the range-ids segmented by their part-id.

(I know that the branch name is wrong...)

### CODE REVIEW

File: core/distributed/partition.cpp [hunk 1]
```diff
@@ -126,6 +128,13 @@ void Partition<LocalIndexType, GlobalIndexType>::finalize_construction()
         get_num_parts(), num_empty_parts_, starting_indices_.get_data(),
         part_sizes_.get_data()));
     size_ = get_element(offsets_, get_num_ranges());
+    array<size_type> range_ids(exec);
+    array<int64> num_ranges_per_part(exec);
+    exec->run(partition::make_build_ranges_by_part(
+        part_ids_.get_const_data(), get_num_ranges(), get_num_parts(),
+        range_ids, num_ranges_per_part));
+    ranges_by_part_ = segmented_array<size_type>::create_from_sizes(
+        std::move(range_ids), num_ranges_per_part);
```
Comment (yhmtsai) [inline core/distributed/partition.cpp:137]: currently, this rely on the atomic add to count the size.
then segmented_array will convert size to offset.
Maybe you can implement a similar kernel with convert_idxs_to_ptrs + additional index map from i to range_id[i] because `range_part[range_id[i]]` is sorted. then segmented_array can `create_from_offsets` directly

File: reference/distributed/partition_helpers.hpp [hunk 2]
```diff
@@ -48,6 +51,49 @@ LocalIndexType map_to_local(
...
+        });
+    auto local_range_id = std::distance(local_ranges.begin, it) - 1;
+    return local_range_id;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+GlobalIndexType map_to_global(
+    LocalIndexType idx,
+    device_partition<const LocalIndexType, const GlobalIndexType> partition,
```
Comment (yhmtsai) [inline reference/distributed/partition_helpers.hpp:87]:
```suggestion
    const device_partition<const LocalIndexType, const GlobalIndexType>& partition,
```

File: reference/distributed/partition_helpers.hpp [hunk 3]
```diff
@@ -48,6 +51,49 @@ LocalIndexType map_to_local(
...
+    size_type range_id)
+{
+    auto range_bounds = partition.offsets_begin;
```
Comment (yhmtsai) [inline reference/distributed/partition_helpers.hpp:90]: assert range_id is lower than num_range?

**Review by yhmtsai — COMMENTED** [review]
not finish review yet.
why is it required for PGM?
after merging row and column, do they still need to map back to the global index?
you need to reindex all of them because the size will be the smaller than the original one after coarsening.

**Comment by MarcelKoch** [comment]
> after merging row and column, do they still need to map back to the global index? 
> you need to reindex all of them because the size will be the smaller than the original one after coarsening.

@yhmtsai I don't understand your point, could you rephrase this?

**Comment by MarcelKoch** [comment]
@yhmtsai I think this discussion makes more sense in #1639. I don't see much relevance to this PR.

**Comment by yhmtsai** [comment]
yes, but if PGM does not need it, do you still need this function for something else?

**Comment by MarcelKoch** [comment]
I think I need at least parts of it to keep the old distributed matrix constructor around. @fritzgoebel also seems to need it for his domain decomposition matrix.

Anyway, this still sounds like a discussion for #1639.

File: test/distributed/index_map_kernels.cpp [hunk 4]
```diff
@@ -388,3 +406,256 @@ TEST_F(IndexMap, GetLocalWithCombinedIndexSpaceWithInvalidIndexSameAsRef)
 
     GKO_ASSERT_ARRAY_EQ(result, dresult);
 }
+
+
+TEST_F(IndexMap, GetGlobalWithLocalIndexSpaceSameAsRef)
```
Comment (pratikvn) [inline test/distributed/index_map_kernels.cpp:411]: Same here. I think just the test with both invalid and valid indices should be enough because the kernel you are calling is the same, and the code path in both cases is the same except for the bounds check.

File: test/distributed/partition_kernels.cpp [hunk 5]
```diff
@@ -69,6 +69,16 @@ class Partition : public CommonTestFixture {
...
+
+        GKO_ASSERT_ARRAY_EQ(
+            gko::make_const_array_view(
+                this->ref, part->get_num_ranges(),
+                part->get_ranges_by_part().get_const_flat_data()),
+            gko::make_const_array_view(
+                this->exec, dpart->get_num_ranges(),
+                dpart->get_ranges_by_part().get_const_flat_data()));
+        GKO_ASSERT_ARRAY_EQ(part->get_ranges_by_part().get_offsets(),
+                            dpart->get_ranges_by_part().get_offsets())
```
Comment (pratikvn) [inline test/distributed/partition_kernels.cpp:81]: If you add a `GKO_ASSERT_SEGEMENT_ARRAY_EQ` macro from the assert function in the other file, then you can use that here as well.

**Review by pratikvn — COMMENTED** [review]
Some comments, but mostly looks good!

File: core/distributed/device_partition.hpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+/**
+ * Explicitly create a const version of device_segmented_array.
+ *
+ * This is mostly relevant for tests.
+ */
+template <typename LocalIndexType, typename GlobalIndexType>
+constexpr device_partition<const LocalIndexType, const GlobalIndexType>
+to_device_const(
```
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply core/distributed/device_partition.hpp:68]: Yes, right now I only need the `to_device_const` version, as I'm not modifying the partition. So I will keep only that version.

File: reference/distributed/partition_helpers.hpp (hunk 2 above)
Comment (MarcelKoch) [reply reference/distributed/partition_helpers.hpp:87]: the `device_partition` type is meant to be handled as a value type. It only contains trivially copyable members.

File: core/distributed/device_partition.hpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+to_device(
+    const experimental::distributed::Partition<LocalIndexType, GlobalIndexType>*
+        partition)
+{
+    auto num_ranges = partition->get_num_ranges();
+    auto num_parts = partition->get_num_parts();
+    return {num_parts,
+            partition->get_num_empty_parts(),
+            partition->get_size(),
+            partition->get_range_bounds(),
```
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply core/distributed/device_partition.hpp:50]: I think this is a bit of a temporary solution until we fully introduce device-side views for all of our (relevant) types. So I would leave handling that issue until then.

File: reference/test/distributed/index_map_kernels.cpp [hunk 8]
```diff
@@ -195,3 +196,113 @@ TEST_F(IndexMap, CanGetLocalWithCombinedISWithInvalid)
...
+
+    gko::array<global_index_type> expected(ref, {14, 13, 12, 5, 4, 3, 13});
+    GKO_ASSERT_ARRAY_EQ(global_ids, expected);
+}
+
+
+TEST_F(IndexMap, CanGetGlobalWithLocalISWithInvalid)
+{
+    gko::array<global_index_type> global_ids(ref);
+    gko::array<local_index_type> local_ids(ref, {5, 4, 10, 3, 2, 1, 0, 100, 4});
```
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply reference/test/distributed/index_map_kernels.cpp:220]: The global_id=17 is not part of the local index set of rank 1, which is used here. I changed it to rank 0 for this test, and now it also outputs gid=17.

File: core/distributed/index_map_kernels.hpp [hunk 9]
```diff
@@ -72,11 +76,36 @@ namespace kernels {
...
+ * The relevant input parameter from the index map are:
+ *
+ * - partition: the global partition
+ * - remote_global_idxs: the remote global indices, segmented by the owning part
+ *                       ids
+ * - rank: the part id of this process
+ *
+ * Any local index that is not part of the specified index space is mapped to
+ * invalid_index.
+ */
```
Comment (pratikvn) [inline core/distributed/index_map_kernels.hpp:91]: I think this is a nice way to capture the documentation of our kernels (And we have a lot of them). I wonder if it is possible to list all the kernels and capture their documentation in the developer documentation.

File: core/test/utils/assertions.hpp [hunk 10]
```diff
@@ -999,6 +999,38 @@ ::testing::AssertionResult array_equal(const std::string& first_expression,
 }
 
 
+template <typename ValueType>
+::testing::AssertionResult segmented_array_equal(
```
Comment (pratikvn) [inline core/test/utils/assertions.hpp:1003]: Can you also add a small test for this in `core/test/utils/assertions_test.cpp`  ?

File: test/distributed/index_map_kernels.cpp (hunk 4 above)
Comment (pratikvn) [reply test/distributed/index_map_kernels.cpp:409]: I think this is not resolved yet ?

**Review by pratikvn — APPROVED** [review]
A couple more comments.

File: test/distributed/index_map_kernels.cpp [hunk 11]
```diff
@@ -117,6 +117,11 @@ class IndexMap : public CommonTestFixture {
 
     IndexMap()
     {
+        std::random_device rd;
+        std::uniform_int_distribution<> engine_dist(
+            0, std::numeric_limits<int>::max());
```
Comment (yhmtsai) [inline test/distributed/index_map_kernels.cpp:122]: `rd` and `engine_dist` seems to be unused

File: reference/distributed/partition_helpers.hpp [hunk 12]
```diff
@@ -48,6 +51,51 @@ LocalIndexType map_to_local(
...
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+GlobalIndexType map_to_global(
+    LocalIndexType idx,
+    device_partition<const LocalIndexType, const GlobalIndexType> partition,
+    size_type range_id)
+{
+    assert(range_id <
+           std::distance(partition.offsets_begin, partition.offsets_end) - 1);
```
Comment (yhmtsai) [inline reference/distributed/partition_helpers.hpp:91]:
```suggestion
    assert(range_id <
           std::distance(partition.offsets_begin, partition.offsets_end));
```
do you need -1?

File: reference/test/distributed/index_map_kernels.cpp [hunk 13]
```diff
@@ -195,3 +196,108 @@ TEST_F(IndexMap, CanGetLocalWithCombinedISWithInvalid)
     gko::array<local_index_type> expected(ref, {2, 3, 0, 1, 2, 4, -1, 1});
     GKO_ASSERT_ARRAY_EQ(local_ids, expected);
 }
+
+
+TEST_F(IndexMap, CanGetGlobalWithLocalISWithInvalid)
```
Comment (yhmtsai) [inline reference/test/distributed/index_map_kernels.cpp:201]:
```suggestion
TEST_F(IndexMap, CanGetGlobalWithLocalIsWithInvalid)
```
Is it supposted to be `IS`?

File: common/cuda_hip/distributed/index_map_kernels.cpp [hunk 14]
```diff
@@ -296,6 +296,90 @@ GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
...
+
+        auto local_ranges = ranges_by_part.get_segment(rank);
+        auto local_ranges_size =
+            static_cast<int64>(local_ranges.end - local_ranges.begin);
+
+        auto it = binary_search(int64(0), local_ranges_size, [=](const auto i) {
+            return starting_indices[local_ranges.begin[i]] >= lid;
+        });
+        auto local_range_id =
+            it != local_ranges_size ? it : max(int64(0), it - 1);
```
Comment (yhmtsai) [inline common/cuda_hip/distributed/index_map_kernels.cpp:337]: will this happen? it seems to be broken if we reach the else branch

File: common/cuda_hip/distributed/partition_kernels.cpp [hunk 15]
```diff
@@ -132,6 +135,36 @@ GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
...
+    thrust::sequence(policy, range_ids_ptr, range_ids_ptr + num_ranges);
+
+    // mutable copy of range_parts such that it can be used as keys for sorting
+    array<int> range_parts_copy{exec, num_ranges};
+    thrust::copy_n(policy, range_parts, num_ranges,
+                   range_parts_copy.get_data());
+    auto range_parts_ptr = range_parts_copy.get_data();
+
+    thrust::stable_sort_by_key(policy, range_parts_ptr,
+                               range_parts_ptr + num_ranges, range_ids_ptr);
```
Comment (yhmtsai) [inline common/cuda_hip/distributed/partition_kernels.cpp:156]: it uses the truth that range ids is also sorted, right?
such that it is sorted by (part_id, range_id)

**Review by yhmtsai — CHANGES_REQUESTED** [review]
the new testing method `GKO_ASSERT_SEGMENTED_ARRAY_EQ` needs a test in core/test/utils/assertions_test.cpp as @pratikvn mentioned
some comments from @pratikvn seems not to be checked yet

File: test/distributed/partition_kernels.cpp [hunk 16]
```diff
@@ -69,6 +69,9 @@ class Partition : public CommonTestFixture {
             gko::make_array_view(
                 this->exec, dpart->get_num_parts(),
                 const_cast<local_index_type*>(dpart->get_part_sizes())));
+
+        GKO_ASSERT_SEGMENTED_ARRAY_EQ(part->get_ranges_by_part(),
```
Comment (upsj) [inline test/distributed/partition_kernels.cpp:73]: Side-node: Since this is a function returning `void`, this will not prevent subsequent operations from executing.

File: reference/distributed/index_map_kernels.cpp [hunk 17]
```diff
@@ -199,6 +199,77 @@ GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
...
 
+template <typename LocalIndexType, typename GlobalIndexType>
+void map_to_global(
+    std::shared_ptr<const DefaultExecutor> exec,
+    device_partition<const LocalIndexType, const GlobalIndexType> partition,
+    device_segmented_array<const GlobalIndexType> remote_global_idxs,
+    experimental::distributed::comm_index_type rank,
+    const array<LocalIndexType>& local_ids,
+    experimental::distributed::index_space is,
+    array<GlobalIndexType>& global_ids)
```
Comment (upsj) [inline reference/distributed/index_map_kernels.cpp:210]: idxs and ids seems to be used interchangably here?

File: reference/distributed/index_map_kernels.cpp [hunk 18]
```diff
@@ -199,6 +199,77 @@ GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
...
+            auto lid = local_ids.get_const_data()[i];
+            global_ids.get_data()[i] = map_non_local(lid);
+        }
+    }
+    if (is == experimental::distributed::index_space::combined) {
+        for (size_type i = 0; i < local_ids.get_size(); ++i) {
+            auto lid = local_ids.get_const_data()[i];
+            global_ids.get_data()[i] = map_combined(lid);
+        }
+    }
```
Comment (upsj) [inline reference/distributed/index_map_kernels.cpp:266]: I think this might be more readable when pulling in the condition `is == ...`, the compiler can hoist this condition out when optimizing.

File: reference/distributed/index_map_kernels.cpp [hunk 19]
```diff
@@ -199,6 +199,77 @@ GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
...
+    device_segmented_array<const GlobalIndexType> remote_global_idxs,
+    experimental::distributed::comm_index_type rank,
+    const array<LocalIndexType>& local_ids,
+    experimental::distributed::index_space is,
+    array<GlobalIndexType>& global_ids)
+{
+    const auto& ranges_by_part = partition.ranges_by_part;
+    auto local_ranges = ranges_by_part.get_segment(rank);
+
+    global_ids.resize_and_reset(local_ids.get_size());
```
Comment (upsj) [inline reference/distributed/index_map_kernels.cpp:215]: local_ids.get_size() gets used multiple times, might be worth a variable?

File: omp/distributed/index_map_kernels.cpp [hunk 20]
```diff
@@ -239,6 +239,77 @@ GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
...
+    auto local_ranges = ranges_by_part.get_segment(rank);
+
+    global_ids.resize_and_reset(local_ids.get_size());
+
+    auto local_size =
+        static_cast<LocalIndexType>(partition.part_sizes_begin[rank]);
+    auto remote_size = static_cast<LocalIndexType>(
+        remote_global_idxs.flat_end - remote_global_idxs.flat_begin);
+    size_type local_range_id = 0;
+    if (is == experimental::distributed::index_space::local) {
```
Comment (upsj) [inline omp/distributed/index_map_kernels.cpp:262]: same here, I think this might be more readable pulling the condition in? Also can we copy the approach from `reference`, since it uses named lambdas instead of unnamed pieces of code

File: dpcpp/distributed/partition_kernels.dp.cpp [hunk 21]
```diff
@@ -130,6 +131,35 @@ GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
...
+                          array<int64>& sizes)
+{
+    auto policy = onedpl_policy(exec);
+
+    range_ids.resize_and_reset(num_ranges);
+    auto range_ids_ptr = range_ids.get_data();
+    // fill range_ids with 0,...,num_ranges - 1
+    run_kernel(
+        exec, [] GKO_KERNEL(auto i, auto rid) { rid[i] = i; }, num_ranges,
+        range_ids_ptr);
```
Comment (upsj) [inline dpcpp/distributed/partition_kernels.dp.cpp:146]: `components::fill_seq_array`?

File: core/test/utils/assertions.hpp [hunk 22]
```diff
@@ -999,6 +999,38 @@ ::testing::AssertionResult array_equal(const std::string& first_expression,
...
+    if (buffer_result == ::testing::AssertionFailure()) {
+        return buffer_result << "Buffers of the segmented arrays mismatch";
+    }
+
+    auto offsets_result =
+        array_equal(first_expression, second_expression, first.get_offsets(),
+                    second.get_offsets());
+    if (offsets_result == ::testing::AssertionFailure()) {
+        return offsets_result << "Offsets of the segmented arrays mismatch";
+    }
```
Comment (upsj) [inline core/test/utils/assertions.hpp:1028]: I would probably check offsets first, since if they mismatch, the rest of the contents make even less sense.

File: core/distributed/index_map_kernels.hpp (hunk 9 above)
Comment (upsj) [reply core/distributed/index_map_kernels.hpp:91]: I agree with Marcel's comment yesterday that Ginkgo developers would rather use code as documentation than the Doxygen, I think we should focus on writing good comments, regardless of whether they work precisely as Doxygen comments.

File: core/distributed/device_partition.hpp [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+    device_segmented_array<const size_type> ranges_by_part;
+};
+
+
+/**
+ * Explicitly create a const version of device_partition.
+ */
+template <typename LocalIndexType, typename GlobalIndexType>
+constexpr device_partition<const LocalIndexType, const GlobalIndexType>
+to_device_const(
```
Comment (upsj) [inline core/distributed/device_partition.hpp:40]: I like this approach!

File: core/distributed/device_partition.hpp [hunk 24]
```diff
@@ -0,0 +1,61 @@
+// SPDX-FileCopyrightText: 2024 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/distributed/device_partition.hpp:5]: There are only very obscure edge cases where pragma once doesn't work properly (because some filesystems don't make it easy to see if two files have the same identity), I think we could safely move from header guards to `#pragma once` if we wanted to. Every relevant compiler supports it.

**Review by upsj — CHANGES_REQUESTED** [review]
Some minor change requests and comments

File: reference/distributed/partition_helpers.hpp (hunk 12 above)
Comment (MarcelKoch) [reply reference/distributed/partition_helpers.hpp:91]: Yes, the offsets array has `#range-ids + 1` entries, since it's an offset array. For an empty partition, `offsets = {0}`, so the begin and end pointers are different.

File: reference/test/distributed/index_map_kernels.cpp (hunk 13 above)
Comment (MarcelKoch) [reply reference/test/distributed/index_map_kernels.cpp:201]: I think I'm usually using `IS` in these tests to denote the `index_space` enum. I would keep it like that.

File: common/cuda_hip/distributed/partition_kernels.cpp (hunk 15 above)
Comment (MarcelKoch) [reply common/cuda_hip/distributed/partition_kernels.cpp:156]: Yes, although the `part_id` is not used after this anymore. It's only relevant that the `range_id` is first sorted by `part_id`, and then by `range_id`.

File: test/distributed/partition_kernels.cpp (hunk 16 above)
Comment (MarcelKoch) [reply test/distributed/partition_kernels.cpp:73]: This should be implemented the same way as the `GKO_ASSERT_ARRAY_EQ` macro. The `segmented_array_equal` function will return something, so I don't know if there is anything more to do here.

Comment (upsj) [reply test/distributed/partition_kernels.cpp:73]: I mean the overall approach of implementing a custom `assert_something` function may cause tests to continue executing after the assertion fails, which may or may not be desirable. It's fine here, but something to keep in mind. To fix this, we would need to return `testing::AssertionResult` and wrap the call in a macro. But nothing to hold back the code here, just a comment that people should be aware of.

File: reference/distributed/index_map_kernels.cpp (hunk 17 above)
Comment (MarcelKoch) [reply reference/distributed/index_map_kernels.cpp:210]: I will now use `idxs` here.

File: reference/distributed/index_map_kernels.cpp (hunk 19 above)
Comment (MarcelKoch) [reply reference/distributed/index_map_kernels.cpp:215]: by hoisting the ifs into the loop this reduced to two times, so I will just leave it.

File: omp/distributed/index_map_kernels.cpp (hunk 20 above)
Comment (MarcelKoch) [reply omp/distributed/index_map_kernels.cpp:262]: I think one compiler didn't work with lambdas within the parallel for loop, maybe nvhpc. I will still hoist the ifs into the loop.

File: core/distributed/partition.cpp (hunk 1 above)
Comment (MarcelKoch) [reply core/distributed/partition.cpp:137]: Yes, for now I would just use atomics. I don't think this is a performance critical part, so it should be fine for a while.

File: omp/distributed/index_map_kernels.cpp (hunk 20 above)
Comment (upsj) [reply omp/distributed/index_map_kernels.cpp:262]: Intel's classic compiler used to have the opposite issue (you needed to put the loop body into an IIFE to work around some compiler crashes). Then maybe the code could use some comments describing the different cases briefly

File: omp/distributed/index_map_kernels.cpp [hunk 25]
```diff
@@ -19,8 +19,6 @@ namespace gko {
 namespace kernels {
 namespace omp {
 namespace index_map {
-
-
```
Comment (yhmtsai) [inline omp/distributed/index_map_kernels.cpp:23]: should be kept

File: reference/distributed/index_map_kernels.cpp [hunk 26]
```diff
@@ -15,8 +15,6 @@ namespace gko {
 namespace kernels {
 namespace reference {
 namespace index_map {
-
-
 template <typename LocalIndexType, typename GlobalIndexType>
```
Comment (yhmtsai) [inline reference/distributed/index_map_kernels.cpp:18]: should kept

**Review by yhmtsai — APPROVED** [review]

**Review by pratikvn — APPROVED** [review]
LGTM!

**Review by upsj — APPROVED** [review]
LGTM!
